pkg/sql/alter_column_type.go

```diff
@@ -0,0 +1,225 @@
+		col.Type = *typ
+	case schemachange.ColumnConversionGeneral:
+		// Make sure that all nodes in the cluster are able to perform general alter column type conversions.
+		if !params.p.ExecCfg().Settings.Version.IsActive(params.ctx, clusterversion.VersionAlterColumnTypeGeneral) {
+			return pgerror.Newf(pgcode.FeatureNotSupported,
+				"all nodes are not the correct version for alter column type general")
+		}
+		if !params.SessionData().AlterColumnTypeGeneral {
+			return pgerror.Newf(pgcode.FeatureNotSupported,
+				"alter column type general is experimental; "+
+					"you can enable alter column type general support by running `SET enable_alter_column_type_general = true")
+		}
+		if err := alterColumnTypeGeneral(tableDesc, col, t); err != nil {
+			return err
+		}
+		if err := params.p.createOrUpdateSchemaChangeJob(params.ctx, tableDesc, "alter column type", sqlbase.InvalidMutationID); err != nil {
+			return err
+		}
+		params.p.SendClientNotice(params.ctx, pgerror.Noticef("alter column type changes are finalized asynchronosuly;"+
+			"further schema changes on this table may be restricted until the job completes; "+
+			"some inserts into the altered column may not be supported until the schema change is finalized"))
+	default:
+		return fmt.Errorf("unknown conversion for %s -> %s",
+			col.Type.SQLString(), typ.SQLString())
+	}
+
+	return nil
+}
+
+func alterColumnTypeGeneral(
+	tableDesc *sqlbase.MutableTableDescriptor,
+	col *sqlbase.ColumnDescriptor,
+	t *tree.AlterTableAlterColumnType,
+) error {
+
+	usingExpressionNotSupportedErr := unimplemented.NewWithIssuef(
+		47706, "alter column type using expression is not supported")
+
+	colInIndexNotSupportedErr := unimplemented.NewWithIssuef(
+		47636, "alter column type requiring rewrite of on-disk "+
+			"data is currently not supported for columns that are part of an index")
+
+	if t.Using != nil {
+		return usingExpressionNotSupportedErr
+	}
+
+	for i := range tableDesc.Indexes {
+		for _, id := range append(
+			tableDesc.Indexes[i].ColumnIDs,
+			tableDesc.Indexes[i].ExtraColumnIDs...) {
+			if col.ID == id {
+				return colInIndexNotSupportedErr
+			}
+		}
+	}
+
+	for _, id := range append(
+		tableDesc.PrimaryIndex.ColumnIDs,
+		tableDesc.PrimaryIndex.ExtraColumnIDs...) {
+		if col.ID == id {
+			return colInIndexNotSupportedErr
+		}
+	}
+
+	currentMutationID := tableDesc.ClusterVersion.NextMutationID
+	for i := range tableDesc.Mutations {
+		mut := &tableDesc.Mutations[i]
+		if mut.MutationID < currentMutationID {
+			return unimplemented.NewWithIssuef(
+				47137, "table %s is currently undergoing a schema change", tableDesc.Name)
+		}
+	}
+
+	nameExists := func(name string) bool {
+		_, _, err := tableDesc.FindColumnByName(tree.Name(name))
+		return err == nil
+	}
+
+	shadowColName := sqlbase.GenerateUniqueConstraintName(
+		col.Name,
+		nameExists,
+	)
+
+	// Compute Expr.
+	unresolvedName := func(name string) *tree.UnresolvedName {
+		return &tree.UnresolvedName{
```

Comment:
Apparently the thing to do is make a `tree.ColumnItem` rather than an unresolved name